docs: adding a storage concepts page (#9155)
hashicorp/vault: website/pages/docs/concepts/storage.mdx

---
layout: docs
page_title: Storage
sidebar_title: Storage
description: >-
  Vault relies on external storage to save its information.
---

# Storage

As described on our [Architecture](/docs/internals/architecture) page, Vault's storage backend is untrusted storage used purely to keep encrypted information.

## Supported Storage Backends

For enterprise customers, HashiCorp provides official support for Consul and Vault's Integrated Storage as storage backends.

Many other options for storage are available with community support - see our [Storage Configuration](/docs/configuration/storage) section for more information.

## Backups & Restores

> Due to the highly flexible nature of Vault's potential storage configurations, providing exacting guidance on backing up and restoring Vault is challenging.

Reviewer: ```suggestion
Due to the highly flexible nature of Vault's potential storage configurations, providing exact guidance on backing up and restoring Vault is challenging.
```

When backing up Vault, there are two pieces to consider:

1. Vault's encrypted data in the storage backend

> 2. Configuration files and management scripts for running the Vault server

Reviewer 2: What scripts are we referring here? The customer written scripts?
If yes, they will likely not be in Vault's store; and do we really want to warn about that?. It might make readers wonder about a storage related artifact to back up which doesn't exist.

Author: Yes - these are things that the customer might use to interact with Vault that are not stored within Vault itself. I do think it's valuable to mention both the desire to save things internal to Vault and things external to Vault as a reminder to operators for when they're coming up plans - like a DR plan, or what needs to be on a new node when it's created and added to a cluster.

So, I want to include both items, but if you have suggestions for how to rephrase the second one to better represent "external-to-Vault stuff that you as an operator might want to save", I'm happy to rephrase?

Reviewer 3: I can see both points of view here.  Maybe some like this for 2

"Disk backup for the underlying server running Vault.  _Note: This would include Vault's configuration file and any custom scripts written to help with the management of Vault and/or the server itself_"

Reviewer 4: Do we also want to mention here, or somewhere below, that the backups will only be able to be decrypted by the seal that is in use at the time of backup?

### Vault's Persisted Data

> In an ideal world, backups and restores would be performed while Vault is offline. If offline backups are impossible, we have recommended using a storage backend that supports atomic snapshots (such as [Consul](https://www.consul.io/docs/commands/snapshot.html) or [Integrated Storage](https://www.vaultproject.io/docs/commands/operator/raft#snapshot)).

Reviewer 5: "have recommended" sounds weird here, why not just "we recommend"?

Author: I'm attempting to further communicate that we recommend it with slightly less preference than taking a backup while Vault is offline. It is a bit awkward, I would prefer to keep it however.

Reviewer: I think Nick's suggestion here is not to remove this, but to slightly re-word it.

> In an ideal world, backups and restores would be performed while Vault is offline. If offline backups are impossible, we have recommended using a storage backend that supports atomic snapshots (such as [Consul](https://www.consul.io/docs/commands/snapshot.html) or [Integrated Storage](https://www.vaultproject.io/docs/commands/operator/raft#snapshot)).

Reviewer 6: ```suggestion
Backups and restores are ideally performed while Vault is offline. If offline backups are not feasible, we recommend using a storage backend that supports atomic snapshots (such as [Consul](https://www.consul.io/docs/commands/snapshot.html) or [Integrated Storage](https://www.vaultproject.io/docs/commands/operator/raft#snapshot)).
```

Reviewer 7: This is tough for 24/7 production/enterprise usage, maybe the guidance is always run a backend that is atomic snapshottable (ie, Consul + IS). And, if not, we rec they're performed offline.

**If your storage backend does not support atomic snapshots, we recommend only taking offline backups.**

> We recommend taking backups **before**, but not during, write operations to the `/sys` API (excluding the `/sys/leases`, `/sys/namespaces`, `/sys/tools`, `/sys/wrapping`, `/sys/policies`, and `/sys/pprof` endpoints). Some examples of workflows that write to the `/sys` API are upgrades, server moves, and rekeys. In the future, this guidance may change for the Integrated Storage backend.

Reviewer: This bit is somewhat hard to enforce on production environments. Spitballing here, but I wonder if we could/should introduce a "maintenance mode" that essentially puts a cluster to be in read-only mode.

Author: I know it's hard to enforce, but I think the guidance is still worthwhile and that any maintenance mode conversation would be separate. Let me know if you disagree on this though.

Reviewer: Yep, this is a good heads-up to keep anyways.

> To perform a backup or restore of Vault's encrypted data, refer to the snapshotting documentation for your storage backend.

Reviewer 5: Can we add links for Raft and Consul snapshot APIs or commands?  This sentence implies that all storage backends support snapshotting, I'd rather we point to how to do it for consul/raft and leave out the word "snapshot" for other backends.

Author: Yup - I'll poke at this paragraph a little more.

> ### Configuration

Reviewer 5: Another suggested thing to backup: any user-installed plugins.

Should we put in a warning here about backing up sensitive configuration?  You might have a vault token for transit autounseal in your config, or you might have a TLS private key in your config dir, etc.

Author: Someone would have to help me write that part, but I think it would be a great addition.

Reviewer 8: > Should we put in a warning here about backing up sensitive configuration? 

Yes I would say that would be best  =)

Author: @ncabatoff would you take another look at this paragraph sometime? If you want to do significant workshopping, let's chat on Slack or Zoom though.

Reviewer 5: LGTM

sidebar_title: Storage
description: >-
  Vault relies on external storage to save its information.
---

# Storage

As described on our [Architecture](/docs/internals/architecture) page, Vault's storage backend is untrusted storage used purely to keep encrypted information.

## Supported Storage Backends

For enterprise customers, HashiCorp provides official support for Consul and Vault's Integrated Storage as storage backends.

Many other options for storage are available with community support - see our [Storage Configuration](/docs/configuration/storage) section for more information.

## Backups & Restores

Due to the highly flexible nature of Vault's potential storage configurations, providing exact guidance on backing up and restoring Vault is challenging.

When backing up Vault, there are two pieces to consider:

1. Vault's encrypted data in the storage backend
2. Configuration files and management scripts for running the Vault server

There's also a big question - what is the error case you're trying to guard against by saving a backup?

### The Big Question - Why Take Backups?

It's important to consider the question of "why take a backup" while developing your ongoing

> backup strategy.

Reviewer 4: This section mostly talks about when not to use backups, should we add a bit more around when backups are useful? i.e. before an upgrade or major change to the cluster, to combat accidental data deletions, etc?

Author: @briankassouf I moved some things around - could you re-read the "Why take Backups" section and see what you think?

> We do not recommend backups as protection against the failure of an individual machine or data

Reviewer 4: I think this is what you're getting at but could be be explicit and say that backups are not a replacement for running in HA?

> center. Vault servers can run in clusters. Vault Enterprise supports replicated clusters for

Reviewer 4: For OSS I do think backups cover the loss of a data center

performance and disaster recovery. With open source features, a Vault cluster can extend across
multiple availability zones within a region.

As you develop a plan for backing up and restoring Vault, include clustering in your plan to
address redundancy concerns.

### Vault's Persisted Data

Backups and restores are ideally performed while Vault is offline. If offline backups are not
feasible, we recommend using a storage backend that supports atomic snapshots (such as
[Consul](https://www.consul.io/docs/commands/snapshot.html) or [Integrated
Storage](/docs/commands/operator/raft#snapshot)).

**If your storage backend does not support atomic snapshots, we recommend only taking offline backups.**

> We recommend taking backups **before**, but not during, write operations to the `/sys` API

Reviewer 5: Do we want to give any explanation for why we recommend this?  e.g. "Even with atomic snapshots, certain Vault operations may consist of multiple writes, and a snapshot taken in the middle of them may result in an inconsistent state and thus a less useful backup."

Author: I followed up with Nick and will follow up with Mark, but the short answer is no. Feel free to reach out to me for further discussion.

Reviewer 4: This section may fit better in the "Why take backups" section.

Author: I think understanding when to take a backup is more of a "how do I" and less of a "should I" kind of thing - could you elaborate on the connection to the "should I" section?

Reviewer 4: I more read it as you "should" take a backup **before** doing system changing operations. Which i think is more useful than saying don't take a snapshot **during** these operations.

(excluding the `/sys/leases`, `/sys/namespaces`, `/sys/tools`, `/sys/wrapping`, `/sys/policies`, and
`/sys/pprof` endpoints). Some examples of workflows that write to the `/sys` API are upgrades,

> server moves, and rekeys. In the future, this guidance may change for the Integrated Storage

Reviewer: Is this referring to replication state changes, or does it entail more than that?

Reviewer 5: I read it as changes to raft peers.

Author: I believe I copied it from someone's suggestion (which may have been on the original google doc), and so am not certain what it means and will remove it.

backend.

To perform a backup or restore of Vault's encrypted data when using a HashiCorp-supported storage
backend, see the instructions linked below. For other storage backends, follow the documentation of
that backend for taking and restoring backups.

* Integrated Storage [snapshots](/docs/commands/operator/raft#snapshot)
* Consul [snapshots](https://www.consul.io/docs/commands/snapshot.html)

### Configuration

In addition to backing up Vault's encrypted data via the storage backend, you may also wish to
save the server configuration files, any scripts for managing the Vault service, and ensure you
can reinstall any user-installed plugins. The location of these files will be specific to your
installation of Vault.

> ~> **NOTE**: Some of your configuration may be sensitive (a Vault token for Transit Autounseal

Reviewer 5: Suggest we add something after this to the effect that snapshots are encrypted and thus do not share this concern.
